go-controller/pkg/clustermanager/healthcheck/controller.go

```diff
@@ -0,0 +1,426 @@
+	if updated {
+		c.informConsumers(task.node)
+	}
+
+	// if no interest from consumers, reconcile to cancel the probing task
+	if !c.hasInterestedConsumers(task.node, nil) {
+		c.controller.Reconcile(task.node)
+	}
+}
+
+func (c *controller) reconcile(nodeName string) error {
+	start := time.Now()
+	klog.V(5).Infof("Reconciling health check for node %s", nodeName)
+	defer func() {
+		klog.V(5).Infof("Finished reconciling health check for node %s: %v", nodeName, time.Since(start))
+	}()
+
+	node, err := c.nodeInformer.Lister().Get(nodeName)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return err
+	}
+
+	if node == nil {
+		// node deleted, stop probing
+		updated := c.cancelProbeWithState(nodeName, UNKNOWN)
+		if updated {
+			klog.Infof("Stopping health state probing for node %s: node deleted", nodeName)
+			c.informConsumers(nodeName)
+		}
+		return nil
+	}
+
+	names := []string{}
+	if !c.hasInterestedConsumers(nodeName, &names) {
+		// no interest, stop probing
+		updated := c.cancelProbeWithState(nodeName, UNKNOWN)
+		if updated {
+			klog.Infof("Stopping health state probing for node %s: no interested consumers out of %q", nodeName, names)
+			// in case someone just registered
+			c.informConsumers(nodeName)
+		}
+		return nil
+	}
+
+	nodeSubnets, err := util.ParseNodeHostSubnetAnnotation(node, types.DefaultNetworkName)
+	if err != nil && !util.IsAnnotationNotSetError(err) {
+		return fmt.Errorf("failed to parse node %s subnets annotation %v", node.Name, err)
+	}
+
+	mgmtIPs := make([]net.IP, len(nodeSubnets))
+	for i, subnet := range nodeSubnets {
+		mgmtIPs[i] = util.GetNodeManagementIfAddr(subnet).IP
+	}
+
+	update := true
+	nodeInfo, probed := c.isProbedWithNodeInfo(nodeName)
+	if probed {
+		less := func(a, b net.IP) bool { return a.String() < b.String() }
+		update = !cmp.Equal(nodeInfo.ips, mgmtIPs, cmpopts.SortSlices(less))
+	}
+
+	if !update {
+		// already probing and no config update, bail out
+		return nil
+	}
+
+	if len(mgmtIPs) == 0 {
+		// no proper config, stop probing and assume unreachable out of
+		// precaution
+		updated := c.cancelProbeWithState(nodeName, UNREACHABLE)
+		if updated {
+			klog.Infof("Stopping health state probing for node %s: node management IPs unknown", nodeName)
+			c.informConsumers(nodeName)
+		}
+		return nil
+	}
+
+	nodeInfo.ips = mgmtIPs
+
+	// add/update probing task
+	klog.Infof("Starting or updating health state probing for node %s on IPs %s", nodeName, util.StringSlice(mgmtIPs))
+	c.startProbeWithNodeInfo(nodeName, nodeInfo)
+
+	return nil
+}
+
+func (c *controller) informConsumers(node string) {
+	c.consumerLock.RLock()
+	defer c.consumerLock.RUnlock()
+	// note we don't query the interest of the consumers as that might be racey,
+	// it is up to the consumers to appropriately filter out uninteresting
+	// events
+	for _, consumer := range c.consumers {
+		consumer.HealthStateChanged(node)
+	}
+}
+
+func (c *controller) hasInterestedConsumers(node string, names *[]string) bool {
```

Comment:
needs doc text, what is names supposed to be? why is this a pointer?

Reply:
Will add. Basically it is for `hasInterestedConsumers` method to fill back the consumer names that were queried so that the caller can log it.

Reply:
it might be cleaner to return a slice,bool:
```
func (c *controller) hasInterestedConsumers(node string) ([]string, bool) {
```

rather than passing pointers to slices

Reply:
I guess I was subjectively interested in keeping a predicate form. I thought being an unexported private method, it wouldn't be that relevant. I will add doc but if you insist I will change it.